cider/exec/plan/validator/CiderPlanValidator.cpp

```diff
@@ -0,0 +1,134 @@
+
+bool CiderPlanValidator::isSupported(const substrait::Plan& plan,
+                                     const std::string& from_platform) {
+  // Assumed that plan pattern already validated, do function look up first
+  // Apply specific rule check on the plan
+  return isSupportedSlice(constructPlanSlice(plan, from_platform), from_platform);
+}
+
+PlanSlice CiderPlanValidator::constructPlanSlice(const substrait::Plan& plan,
+                                                 const std::string& from_platform) {
+  const substrait::Rel& root_rel = getRootRel(plan);
+  std::vector<substrait::Rel> rel_vec;
+  putRelNodesInVec(root_rel, rel_vec);
+  return PlanSlice{rel_vec, from_platform};
+}
+
+bool CiderPlanValidator::isSupportedSlice(const PlanSlice& plan_slice,
+                                          const std::string& from_platform) {
+  // TODO: (yma11) add function look up and rule based check
+  return true;
+}
+
+void CiderPlanValidator::putRelNodesInVec(const substrait::Rel& rel_node,
+                                          std::vector<substrait::Rel>& rel_vec) {
+  const substrait::Rel::RelTypeCase& rel_type = rel_node.rel_type_case();
+  switch (rel_type) {
+    case substrait::Rel::RelTypeCase::kRead: {
+      rel_vec.push_back(rel_node);
+      break;
+    }
+    case substrait::Rel::RelTypeCase::kFilter: {
+      rel_vec.push_back(rel_node);
+      putRelNodesInVec(rel_node.filter().input(), rel_vec);
+      break;
+    }
+    case substrait::Rel::RelTypeCase::kProject: {
+      rel_vec.push_back(rel_node);
+      putRelNodesInVec(rel_node.project().input(), rel_vec);
+      break;
+    }
+    case substrait::Rel::RelTypeCase::kAggregate: {
+      rel_vec.push_back(rel_node);
+      putRelNodesInVec(rel_node.aggregate().input(), rel_vec);
+      break;
+    }
+    case substrait::Rel::RelTypeCase::kJoin: {
+      rel_vec.push_back(rel_node);
+      putRelNodesInVec(rel_node.join().left(), rel_vec);
+      break;
+    }
+    default:
+      CIDER_THROW(CiderCompileException,
+                  fmt::format("Unsupported substrait rel type {}", rel_type));
+  }
+}
+
+substrait::Rel CiderPlanValidator::getRootRel(const substrait::Plan& plan) {
+  if (plan.relations_size() == 0) {
+    CIDER_THROW(CiderCompileException, "invalid plan with no root node.");
+  }
+  if (!plan.relations(0).has_root()) {
+    CIDER_THROW(CiderCompileException, "invalid plan with no root node.");
+  }
+  return plan.relations(0).root().input();
+}
+
+PlanSlice CiderPlanValidator::getCiderSupportedSlice(substrait::Plan& plan,
+                                                     std::string& from_platform) {
+  substrait::Rel root = getRootRel(plan);
+  // Put plan rels in a vector with pair <substrait::Rel, isSupported>
+  std::vector<substrait::Rel> rel_vec;
+  putRelNodesInVec(root, rel_vec);
+  // Compose plan slices based on SingleNodeValidator validate result
+  std::vector<PlanSlice> slice_candidates;
+  for (int i = 0; i < rel_vec.size(); i++) {
```

Comment:
after this loop it's possible to have 2 or more sequence slice. if we have a plan like `A->B->C->D->E->F->G`,  it's possible to have `[B->C]`,  `[E->F]` in slice_candidates.

Reply:
Yes. If they have same length, we will return the first one.  We may can add better rule for final slice select later.